# Exploratory Data Analysis of CPI Data
Using Bokeh.

Author: Jayanth Raman, Date: 2019-11-22

# Consumer Price Index (CPI)
A Consumer Price Index (CPI) measures changes in the price level of a weighted average market basket of consumer goods and services purchased by households. [Wikipedia](https://en.wikipedia.org/wiki/Consumer_price_index)
The CPI is a statistical estimate constructed using the prices of a sample of representative items whose prices are collected periodically. Sub-indices and sub-sub-indices are computed for different categories and sub-categories of goods and services.  The overall index is computed as a weighted average of sub-indices for different components of consumer expenditure, such as food, housing, shoes, clothing, each of which is in turn a weighted average of sub-sub-indices.  The annual percentage change in a CPI is used as a measure of inflation.

# Observations (USA)
The USA file contains monthly CPI data for 14 cities from 1976 through mid-1995.  For each city, the data contains the overall CPI for that city and also the CPI for 14 different categories.  Some of the data is missing--more on this later.

The CPI around mid 1983 is around 100.  Hence we could view mid-1983 as a reference point.

The cities are: Baltimore, Boston, Chicago, Dallas, Detroit, Houston, Los Angeles, Miami, New York City, Philadelphia, Pittsburgh, San Francisco, St. Louis, and Washington D.C.

The categories are: Food at home; Food away from home; Alcoholic beverages; Shelter; Fuel and other utilities; Household furnishings & operations; Men's and boy's apparel; Women's and girl's apparel; Footwear; Private transportation; Public transportation; Medical care; Personal care; and Entertainment.

Below, we examine each of the categories.

## 0. City CPI
Among the cities, in the mid-70s, NY has the highest overall CPI
followed by Boston and Philly.  On the lower end are Dallas, Houston,
Pittsburgh, and San Francisco.

In the mid-90s, NY continues to top the charts and Boston and Philly
continue to stay at the top.  San Francisco and Chicago, which had low
CPI in the mid-70s, now have higher CPI comparitively.

What this says is that NY, Boston and Philly have always been
expensive cities to live in, whereas SF and Chicago started off as
affodable cities, but became expensive in the mid-90s.

## 1. Food at home
This category seems to have the most data, surprisingly.

In the mid-70s, Boston is by far the most expensive, followed by
Philly, Baltimore and Chicago.  DC, Dallas, LA and SF start out as the
least expensive.

In the mid-90s, Dallas and Pittsburgh are the least expensive,
followed by Detroit and St. Louis--all of these cities have a CPI
under 144.  Houston, Boston and DC have moderate CPI.  The rest of the
pack have CPIs at or above 155.

## 2. Food away from home
Detroit, DC and Boston are the most expensive in the mid-70s.  Houston
is by far the least expensive, followed by Dallas, Philly and SF.

In the mid-90s, Miami is the most expensive, although data for it is
sparse in the mid-70s, followed by NY and Boston.  Houston remains the
least expensive, followed by Philly.  Dallas has moved to the top tier
whereas SF is moderate.

## 3. Alchoholic beverages
In the mid-70s, Chicago is by far the most expensive followed by NY.
Dallas, Houston and Pittsburgh are the least expensive.

In the mid-90s, LA is the most expensive in this category, followed by
Philly, Dallas and Chicago.  Houston, Baltimore and DC are the least
expensive.

Dallas has moved from being among the least expensive in the mid-70s
to being the most expensive in the mid-90s.

### 1991 Federal tax change
There is a **distinctive jump** (increase) in the CPI between the end of
1990 and the beginning of 1991!  The reason for this is probably the
change in the federal tax on alcohol that took effect on January 1,
1991 [see here](https://www.nber.org/digest/may12/w17709.html).

## 4. Shelter
The overall rate of increase in shelter is higher than that of
inflation.  We note this by comparing the CPI for this category with
the overall CPI.

In the mid-70s, Philly is the most expensive followed by NY.  Chicago
and Dallas are the least expensive, followed by St. Louis and
Pittsburgh.  Philly's CPI in this category is almost 38% more
expensive than Chicago's.

In the mid-90s, NY and Philly continue to be the most expensive,
although they have swapped places.  Houston and Dallas are the least
expensive.  Chicago, which was at the bottom of the pack in the
mid-70s, has now moved to the third position.  The disparity between
the least and the most expensive has held ground and is now 41%.

Due to the ratio of the highest to the lowest remaining constant and
the CPIs generally rising, this leads visually to a **large
variation** among the cities in the mid-90s compared to that in the
mid-70s.

## 5. Fuel and other utilities
The **overall rate** of increase of fuel is below that of inflation.
We note this by comparing the CPI for this category with the overall
CPI.

In the mid-70s, Houston is by far the least expensive, followed by
Detroit and San Francisco.  Dallas, which is not that far from
Houston, is around 11% more expensive.  St. Louis is the most
expensive, followed by NY and LA.  The difference between the least
and most expensive is around 27%.

In the mid-90s, LA and SF are by far the most expensive, while Houston
continues to be the least expensive.  The prices in LA are almost
double that of Houston in the data for April 1995.

### Seasonal changes
**Seasonal changes** are evident in the graph--with prices rising in
the summer and falling in autumn.  This can be seen in the rectangular
wave nature of the graphs in many cities.

Also, it appears that the data for this category is more dense than
that for other categories.

## 6. Household furnishings & operations
Price changes in this category underperform overall inflation.

In the mid-70s, Houston has the highest prices in this category
followed by LA.  Dallas, SF, and Baltimore are the least expensive
along with Philly and Chicago.

In the mid-90s, DC is the most expensive, followed by Dallas, Miami
and NY.  Detroit is the least expensive, followed by Houston (which
used to be the most expensive).

## 7. Men's and boy's apparel
The prices in the category have not changed much over the years
compared to the other categories.

Chicago is the most expensive in the mid-70s while Houston is the
least expensive.  They have a differential of 43% between them.

In the mid-90s, Houston is now the most expensive and shows large
variations in prices--in early to mid 1993, the CPI crosses 180, but
falls back below 160 in mid 1995.  Philly and SF are the least
expensive.  The differential between the most and least expensive is
54% in early 1995.

## 8. Women's and girl's apparel
The prices in this cateogory have moved less compared to other
categories and even less so compared to men's apparel.

In the mid-70s, Philly and LA are the most expensive--in fact, more
expensive than the mid-1983 probable baseline.  Houston, SF, Boston,
DC and St. Louis are the least expensive.

In the mid-90s, SF and Philly are by far the least expensive--in fact,
less expensive than in mid-1983.  Miami is the most expensive in this
category.  There appears to be a larger variation in the more
expensive cities.  Is this caused by the basket containing items that
vary more, such as designer clothing?

## 9. Footwear
In the mid-70s, Detroit and St. Louis are the most expensive, whereas
Boston is the least expensive followed by Houston and DC.  The
differential between them is around 55%.

In the mid-90s, Chicago is the most expensive, followed by Baltimore,
Boston, Miami and NY.  St. Louis is the least expensive followed by
Philly and LA.  The differential is around 73%.

What part does designer footware and clothing play in the CPI of
footwear and clothing?  Does that account for the higher variability
in some cities?

## 10. Private transportation
A general **downward trend** from 1/1986 through 4/1986 can be observed.
Fuel oil prices dropped 33.4% in 1986 and were 40.4% below their peak
level of April 1981 [see
here](https://books.google.com/books?id=rHamOfrhVt4C&lpg=RA5-PA2&ots=w5L1m2JlUl&f=false).
This is somewhat visible in the fuel and utilities plot, although that
plot also includes utilities.

In 1/1976, there isn't much variability between the cities varying
from 49.1 in Pittsburgh to 58.6 in Boston--for a difference of 19.3%.

In 5/1995, the variability remains about the same--a difference of
17.2% between 124.7 in SF and 146.2 in Philly.

## 11. Public transportation
Public transportation prices rise slowly from 1976 through 1979 and
accelerate their rise in 1980.  The prices between the cities is
bunched together while rising till 1988, when they start to diverge.
The price of public transportation in LA rises faster and higher than
in the other cities.

Prices in 1/1976 hover between 39.7 in Pittsburgh to 58.0 in
Houston--for a difference of 46%.  LA is at 42.6, closer to the lower
end.

In 1995, prices in LA have exceeded 200.  St. Louis has the lowest
price.  The difference between the two is almost 60%.

## 12. Medical care
In 1976, medical care costs are highest in NY while Houston,
Pittsburgh and LA have the lowest costs.  The range is from 45.5 to
54.5 for a difference of almost 20%.

In 1995, costs have skyrocketed to over 200 across all cities.  Costs
in NY have more than quarupled to 226.  The price is highest in Boston
at 265.1 and lowest in Miami at 201.7 for a difference of 31.4%.

Given the skyrocketing healthcare costs in the US, it is not
surprising that this category shows the biggest jump in prices between
1976 and 1995.

## 13. Personal care
In 1976, prices vary from 52.5 in Houston to 71.5 in Detroit.  A
difference of 30.7% can be seen in 10/1976 between the two cities.

In 1995, prices vary from 103.8 in Miami to 189 in Philly.  In 5/1995,
the difference in prices between the two cities is 81.6% which is
rather large.

## 14. Entertainment
In 1976, prices vary from 59.9 in Houston to 74.8 in Detroit.  Prices
in Detroit are around 20% higher than in Houston.

In 1995, prices vary from 140.2 in Miami to 167.1 in Boston--prices in
Boston are 19% higher than in Miami.

Chicago which was at the bottom of the pack in 1976, has moved close
to the top in 1995.  And LA, which was near the top in 1976 has moved
to near the bottom in 1995.

In [1]:
import bokeh.plotting as BP
import bokeh.palettes
import numpy as np
import pandas as pd
# pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 32)
# pd.set_option('display.width', 1000)
BP.reset_output()
# BP.output_notebook()
BP.output_file('prob-hw1-usa.html')

Load the data.

In [2]:
usa = pd.read_excel('USA.xls', na_values=-999.999)
usa.rename({'Unnamed: 0': 'Year-Month'}, axis=1, inplace=True)
canada = pd.read_excel('CAN.xls', na_values=-999.999)
canada.rename({'Unnamed: 0': 'Year-Month'}, axis=1, inplace=True)
# Convert column 0 to a datatime object
usa['Year-Month'] = pd.to_datetime(usa['Year-Month'], format='%Y:%m')
canada['Year-Month'] = pd.to_datetime(canada['Year-Month'], format='%Y:%m')

In [3]:
usa

Year-Month    CH0    CH1    CH2    CH3    CH4    CH5    CH6    CH7    CH8  \
0   1976-01-01   53.7   65.0   58.8   73.3   44.0   47.0   64.6   94.3   96.9   
1   1976-02-01   54.1   65.0   58.9   73.3   44.0   47.6   65.0   95.8   98.5   
2   1976-03-01   54.2   64.3   59.3   73.5   44.0   48.5   65.4   97.1   98.5   
3   1976-04-01   54.5   65.0   59.6   75.2   44.1   48.7   65.6   95.2   98.1   
4   1976-05-01   54.8   64.7   59.8   75.2   44.3   49.1   65.5   97.2   98.3   
5   1976-06-01   55.2   65.4   60.1   75.5   44.4   49.6   66.0   95.9   99.9   
6   1976-07-01   55.5   66.1   60.5   75.5   44.6   50.0   66.0   95.4   97.0   
7   1976-08-01   55.8   66.2   60.7   75.1   45.0   50.0   66.0   94.6  102.8   
8   1976-09-01   56.1   65.7   60.8   75.2   45.2   50.3   66.5   95.9  105.9   
9   1976-10-01   56.3   65.9   61.1   75.1   45.5   49.9   66.8   96.4  103.6   
10  1976-11-01   56.4   65.3   61.2   75.5   45.5   49.9   66.9   96.2  104.3   
11  1976-12-01   56.6   65.1   61.5   75.6   45.9   51.8   67.2   95.9  103.9   
12  1977-01-01   56.7   65.3   62.3   76.0   46.0   52.3   67.3   94.3   96.9   
13  1977-02-01   57.3   66.8   62.6   75.6   46.5   52.6   67.6   95.0   98.9   
14  1977-03-01   57.7   67.4   63.4   75.9   46.3   53.1   68.0   97.2   99.2   
15  1977-04-01   58.2   68.3   64.2   75.7   46.5   54.6   68.5   98.1   98.5   
16  1977-05-01   58.4   68.1   64.8   75.7   46.7   54.7   68.5   99.3   97.3   
17  1977-06-01   58.6   68.9   64.9   76.2   46.7   53.8   69.0   98.9  101.0   
18  1977-07-01   59.0   69.7   65.3   76.0   47.6   54.2   68.9   96.3  101.1   
19  1977-08-01   59.3   70.4   65.1   76.5   48.0   54.3   69.0   98.0  103.2   
20  1977-09-01   59.7   70.8   65.1   77.1   48.7   53.6   69.2  100.0  105.5   
21  1977-10-01   59.8   70.8   65.9   77.5   48.7   54.4   69.6   97.5  103.2   
22  1977-11-01   60.0   70.6   66.0   77.7   49.2   54.6   69.7   98.2  103.7   
23  1977-12-01   60.2   71.3   66.1   78.0   49.2   55.1   69.9   97.8  103.1   
24  1978-01-01   61.1   73.1   68.1   77.8   50.8   55.1   69.8   93.1   98.8   
25  1978-02-01   61.7   74.2   68.8   79.5   51.8   55.5   70.0   92.9   95.8   
26  1978-03-01   62.4   74.8   69.8   80.2   52.9   56.7   71.1   94.8   98.2   
27  1978-04-01   62.7   76.1   70.5   79.6   52.8   57.1   71.9   92.8  103.4   
28  1978-05-01   63.3   77.0   70.6   80.3   53.3   57.5   72.0   94.0  105.1   
29  1978-06-01   63.7   77.9   71.5   81.2   53.8   57.3   72.6   93.3  105.0   
..         ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
203 1992-12-01  142.9  143.3  133.3  153.2  158.1  118.2  118.1  121.9  117.2   
204 1993-01-01  143.2  145.5  133.2  157.6  158.2  118.5  117.5  120.0  114.4   
205 1993-02-01  143.6  142.9  133.9  153.7  158.5  116.9  115.7  125.4  137.9   
206 1993-03-01  144.1  144.7  133.8  157.6  158.1  116.0  116.1  126.3  144.4   
207 1993-04-01  144.7  144.0  134.2  157.8  159.0  118.3  118.6  126.0  136.7   
208 1993-05-01  145.7  146.0  134.9  156.9  159.6  119.5  117.4  125.9  138.3   
209 1993-06-01  145.6  145.0  135.1  156.5  159.6  125.2  118.7  124.6  123.0   
210 1993-07-01  145.5  145.3  135.5  156.1  159.9  125.0  112.6  129.0  121.5   
211 1993-08-01  146.1  145.5  135.3  156.8  160.2  125.8  116.3  125.7  138.3   
212 1993-09-01  146.7  145.8  135.1  158.2  161.2  125.7  117.7  131.9  132.1   
213 1993-10-01  147.2  147.5  135.2  157.7  163.0  122.3  118.0  140.3  126.2   
214 1993-11-01  146.4  147.7  135.5  157.1  163.4  110.9  117.8  139.8  123.7   
215 1993-12-01  146.1  149.9  135.4  157.0  163.6  111.8  116.2  131.7  115.3   
216 1994-01-01  146.5  151.9  135.9  159.2  163.3  111.5  116.2  128.1  119.2   
217 1994-02-01  146.8  150.5  135.8  159.9  163.8  110.8  115.7  120.0  140.6   
218 1994-03-01  147.6  151.4  135.6  161.5  164.2  112.2  115.0  123.5  141.0   
219 1994-04-01  147.9  152.0  136.0  161.0  165.1  111.5  116.0  128.6  128.2   
220 1994-05-01  147.6  150.5  136.9

In [4]:
canada

Year-Month     Q0     M0     O0     T0     W0     R0     E0     C0     V0  \
0   1974-01-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
1   1974-02-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
2   1974-03-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
3   1974-04-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
4   1974-05-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
5   1974-06-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
6   1974-07-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
7   1974-08-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
8   1974-09-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
9   1974-10-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
10  1974-11-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
11  1974-12-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
12  1975-01-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
13  1975-02-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
14  1975-03-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
15  1975-04-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
16  1975-05-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
17  1975-06-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
18  1975-07-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
19  1975-08-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
20  1975-09-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
21  1975-10-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
22  1975-11-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
23  1975-12-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
24  1976-01-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
25  1976-02-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
26  1976-03-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
27  1976-04-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
28  1976-05-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
29  1976-06-01    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
..         ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
227 1992-12-01  128.6  130.4  128.8  130.5  128.5  128.8  127.4  126.8  129.5   
228 1993-01-01  128.9  131.0  129.4  130.8  128.7  129.7  127.8  127.6  130.6   
229 1993-02-01  128.9  131.1  129.7  131.6  129.3  129.7  127.9  127.7  131.0   
230 1993-03-01  129.1  131.3  130.1  131.6  129.6  129.3  126.0  126.1  130.9   
231 1993-04-01  129.0  131.0  130.0  131.3  129.9  131.0  126.6  126.9  131.1   
232 1993-05-01  128.9  131.0  130.2  131.3  130.2  131.5  126.6  127.4  131.8   
233 1993-06-01  129.0  131.0  130.4  131.5  130.1  131.3  127.6  127.8  131.5   
234 1993-07-01  129.2  130.9  130.6  132.0  130.4  131.5  128.1  128.4  132.1   
235 1993-08-01  129.0  130.7  130.7  132.0  130.5  131.8  128.0  128.4  132.6   
236 1993-09-01  129.0  131.0  130.8  132.2  130.5  131.9  127.9  128.9  132.9   
237 1993-10-01  129.2  130.8  130.9  132.4  131.0  132.2  128.0  128.4  132.9   
238 1993-11-01  130.2  132.0  131.3  132.7  131.7  132.7  128.5  129.1  133.5   
239 1993-12-01  129.7  131.8  131.4  132.6  131.7  132.8  128.7  129.0  133.3   
240 1994-01-01  129.7  131.6  131.1  132.4  131.4  133.2  128.8  128.9  133.9   
241 1994-02-01  127.7  129.1  130.3  131.9  131.4  132.6  128.7  128.8  134.1   
242 1994-03-01  127.8  129.1  130.5  131.5  131.5  132.8  128.2  128.6  134.0   
243 1994-04-01  127.8  129.0  130.7  131.5  131.6  133.4  129.1  128.8  133.7   
244 1994-05-01  127.3  128.3  130.5

In [5]:
# Columns in the datasets
print(', '.join(usa.columns))
print(', '.join(canada.columns))

Year-Month, CH0, CH1, CH2, CH3, CH4, CH5, CH6, CH7, CH8, CH9, CH10, CH11, CH12, CH13, CH14, LA0, LA1, LA2, LA3, LA4, LA5, LA6, LA7, LA8, LA9, LA10, LA11, LA12, LA13, LA14, NY0, NY1, NY2, NY3, NY4, NY5, NY6, NY7, NY8, NY9, NY10, NY11, NY12, NY13, NY14, PH0, PH1, PH2, PH3, PH4, PH5, PH6, PH7, PH8, PH9, PH10, PH11, PH12, PH13, PH14, DA0, DA1, DA2, DA3, DA4, DA5, DA6, DA7, DA8, DA9, DA10, DA11, DA12, DA13, DA14, DT0, DT1, DT2, DT3, DT4, DT5, DT6, DT7, DT8, DT9, DT10, DT11, DT12, DT13, DT14, HS0, HS1, HS2, HS3, HS4, HS5, HS6, HS7, HS8, HS9, HS10, HS11, HS12, HS13, HS14, PI0, PI1, PI2, PI3, PI4, PI5, PI6, PI7, PI8, PI9, PI10, PI11, PI12, PI13, PI14, SF0, SF1, SF2, SF3, SF4, SF5, SF6, SF7, SF8, SF9, SF10, SF11, SF12, SF13, SF14, BA0, BA1, BA2, BA3, BA4, BA5, BA6, BA7, BA8, BA9, BA10, BA11, BA12, BA13, BA14, BO0, BO1, BO2, BO3, BO4, BO5, BO6, BO7, BO8, BO9, BO10, BO11, BO12, BO13, BO14, MI0, MI1, MI2, MI3, MI4, MI5, MI6, MI7, MI8, MI9, MI10, MI11, MI12, MI13, MI14, ST0, ST1, ST2, ST3, ST4, ST5

Observations:
 * USA has 233 rows while Canada has 257 rows.  Each row represents data for one month.  The years represented are 1976 through part of 1995 for USA and 1974 through part of 1995 for Canada.
 * USA has 14 cities.
 * Canada has 9 cities (or regions).
 * There are 16 columns for each city: columns numbered 0 through 14 for CPI data, and a column "W" for wages.

In [6]:
# Labels for cities and columns describing goods.
cities = {}
cities['usa'] = {
"CH": "Chicago",
"LA": "Los Angeles",
"NY": "New York City",
"PH": "Philadelphia",
"DA": "Dallas",
"DT": "Detroit",
"HS": "Houston",
"PI": "Pittsburgh",
"SF": "San Francisco",
"BA": "Baltimore",
"BO": "Boston",
"MI": "Miami",
"ST": "St. Louis",
"WA": "Washington, DC",
}
cities['canada'] = {
    "Q": "Quebec",
    "M": "Montreal",
    "O": "Ottawa",
    "T": "Toronto",
    "W": "Winnipeg",
    "R": "Regina",
    "E": "Edmonton",
    "C": "Calgary",
    "V": "Vancouver"
}
goods_descriptions = {
    0: 'City CPI',
    1: 'Food at home',
    2: 'Food away from home',
    3: 'Alcoholic beverages',
    4: 'Shelter',
    5: 'Fuel and other utilities',
    6: 'Household furnishings & operations',
    7: "Men's and boy's apparel",
    8: "Women's and girl's apparel",
    9: 'Footwear',
    10: 'Private transportation',
    11: 'Public transportation',
    12: 'Medical care',
    13: 'Personal care',
    14: 'Entertainment'
}
len(cities['usa']), len(cities['canada']), len(goods_descriptions)

(14, 9, 15)

In [7]:
cdsusa = BP.ColumnDataSource(usa)
cdscan = BP.ColumnDataSource(canada)

In [8]:
##### Notes:
# p.line() does not plot NaN values.  So, if there are alternate NaN values, then no line is drawn!
# Hence, we also use p.circle() to plot those values.
# Panning is linked.  So, if you pan one plot, it will pan the others as well.  And if you select a box or
# oval area in one plot, it will do the same in the other plots as well.
# To show all the plots, click on the gray area to the left of the plots or
# Cell -> Current/All Outputs -> Toggle Scrolling
p = []
# TOOLS = "pan,box_zoom,reset,save"
TOOLS = "crosshair,pan,box_zoom,reset,box_select,lasso_select,save,tap"
colors = bokeh.palettes.d3['Category20'][20]
for ii in range(len(goods_descriptions)):
    title = f'{ii}. {goods_descriptions[ii]}'
    if ii:
        p.append(BP.figure(width=800, height=400, title=title, x_axis_type="datetime",
                           x_range=p[0].x_range, y_range=p[0].y_range, tools=TOOLS))
    else:
        p.append(BP.figure(width=800, height=400, title=title, x_axis_type="datetime",
                           y_range=(0, 250), tools=TOOLS))
    for jj, city in enumerate(cities['usa']):
        p[ii].line('Year-Month', f'{city}{ii}', source=cdsusa, legend_label=city, line_width=2, color=colors[jj])
        p[ii].circle('Year-Month', f'{city}{ii}', source=cdsusa, legend_label=city, size=2, color=colors[jj])        
    p[ii].legend.orientation = "horizontal"
    p[ii].legend.location = 'bottom_right'
q = BP.gridplot(p, ncols=1)
BP.show(q)

In [9]:
min_year = min(usa['Year-Month']).year
max_year = max(usa['Year-Month']).year
min_year, max_year

(1976, 1995)

In [10]:
usa.dropna()

Empty DataFrame
Columns: [Year-Month, CH0, CH1, CH2, CH3, CH4, CH5, CH6, CH7, CH8, CH9, CH10, CH11, CH12, CH13, CH14, LA0, LA1, LA2, LA3, LA4, LA5, LA6, LA7, LA8, LA9, LA10, LA11, LA12, LA13, LA14, NY0, NY1, NY2, NY3, NY4, NY5, NY6, NY7, NY8, NY9, NY10, NY11, NY12, NY13, NY14, PH0, PH1, PH2, PH3, PH4, PH5, PH6, PH7, PH8, PH9, PH10, PH11, PH12, PH13, PH14, DA0, DA1, DA2, DA3, DA4, DA5, DA6, DA7, DA8, DA9, DA10, DA11, DA12, DA13, DA14, DT0, DT1, DT2, DT3, DT4, DT5, DT6, DT7, DT8, DT9, DT10, DT11, DT12, DT13, DT14, HS0, HS1, HS2, HS3, HS4, HS5, HS6, HS7, HS8, ...]
Index: []

[0 rows x 226 columns]

All rows have at least one null value i.e. none of the rows is complete.

In [11]:
usadesc = usa.describe()
usadesc = usadesc.T
usadesc

count         mean          std      min       25%     50%      75%  \
CH0     233.0   106.370386    29.227897    53.70    86.700  108.50  132.000   
CH1     233.0   110.029185    25.795085    64.30    96.000  103.30  135.200   
CH2     233.0   104.248927    22.750771    58.80    91.600  106.20  123.200   
CH3     233.0   112.962661    26.989484    73.30    93.400  106.40  130.600   
CH4     233.0   110.544206    39.214526    44.00    79.900  114.20  146.600   
CH5     233.0    93.718884    22.518044    47.00    77.600   99.90  111.500   
CH6     233.0    99.670386    16.697368    64.60    88.700  105.30  114.000   
CH7     233.0   107.186695    12.469806    88.30    96.300  104.60  117.600   
CH8     233.0   109.668240    12.111580    86.50   100.500  105.90  118.200   
CH9     233.0   109.457082    20.931974    73.80    94.500  103.60  132.600   
CH10    233.0    99.805150    21.094146    53.60    90.700  103.20  116.600   
CH11    233.0   100.838197    33.553354    43.10    67.800  103.10  128.200   
CH12    233.0   123.141631    51.028257    49.70    77.800  114.60  164.300   
CH13    217.0   109.532258    24.013302    61.30    93.000  111.70  129.100   
CH14    233.0   110.590129    30.909061    60.30    87.500  105.70  139.200   
LA0     233.0   108.613305    30.331889    55.40    86.500  109.60  135.600   
LA1     233.0   107.760515    27.195640    59.00    90.400  104.70  131.300   
LA2     233.0   105.712017    25.783185    56.80    85.000  109.20  129.300   
LA3     233.0   114.969099    30.519721    68.00    91.800  107.30  137.900   
LA4     233.0   113.193562    36.200918    50.30    86.800  114.40  149.500   
LA5     233.0   102.063519    31.064868    50.60    74.600  108.50  126.100   
LA6     233.0   101.574249    16.166241    67.40    92.200  104.20  113.800   
LA7     233.0   108.037768    16.661455    79.70    95.200  105.50  123.500   
LA8     233.0   111.380687     9.440714    91.50   104.100  110.70  117.100   
LA9     233.0    96.687983    14.361771    67.40    90.200  100.40  105.700   
LA10    233.0   101.493562    25.096327    51.80    84.700  105.90  121.600   
LA11    233.0   121.043777    49.830553    42.60    90.400  120.40  168.200   
LA12    233.0   122.622318    52.749634    46.40    75.200  114.60  163.600   
LA13    217.0   114.541935    30.695902    57.80    92.100  114.20  142.900   
LA14    233.0   106.381974    22.257080    68.90    88.700  108.10  124.100   
...       ...          ...          ...      ...       ...     ...      ...   
WA0     113.0   110.281416    28.959379    56.60    89.700  111.50  138.000   
WA1     233.0   107.835622    26.837660    59.50    89.400  106.50  133.600   
WA2     128.0   105.648438    25.849656    62.70    86.100  103.20  130.575   
WA3     113.0   112.499115    28.381201    64.40    90.500  111.20  138.800   
WA4     113.0   114.449558    34.707099    51.80    89.600  116.40  148.800   
WA5     128.0    91.835938    24.221025    48.50    67.125  101.75  109.925   
WA6     113.0   107.741593    21.365212    65.80    92.100  109.50  123.800   
WA7     113.0   113.895575    23.314470    78.20    95.300  108.90  136.900   
WA8     113.0   110.351327    22.438806    78.50    90.000  105.30  130.800   
WA9     113.0   104.267257    21.207076    65.70    89.600  101.60  124.500   
WA10    113.0   103.959292    23.692889    54.60    95.300  103.90  124.800   
WA11    113.0   100.187611    27.792943    45.60    84.100  104.60  117.300   
WA12    113.0   123.838938    48.171971    51.70    78.400  119.40  164.300   
WA13    113.0   113.987611    30.992819    63.90    86.700  109.50  145.300   
WA14    113.0   113.461947    27.911481    67.30    89.000  115.20  137.200   
BAW     231.0    10.084416     2.389495     5.55     8.145   10.27   12.015   
BOW     231.0     9.576104     2.629224     5.09     7.265    9.63   11.870   
DAW     231.0     8.615238     2.033428     4.55     6.970    9.17   10.400   
DTW     231.0    12.793203     2.915968     6.91    10.640

In [12]:
# How many have values for all the months?
usadesc[usadesc['count'] == 233.0]

count        mean        std   min    25%    50%    75%    max
CH0   233.0  106.370386  29.227897  53.7   86.7  108.5  132.0  153.1
CH1   233.0  110.029185  25.795085  64.3   96.0  103.3  135.2  157.5
CH2   233.0  104.248927  22.750771  58.8   91.6  106.2  123.2  139.4
CH3   233.0  112.962661  26.989484  73.3   93.4  106.4  130.6  163.7
CH4   233.0  110.544206  39.214526  44.0   79.9  114.2  146.6  174.2
CH5   233.0   93.718884  22.518044  47.0   77.6   99.9  111.5  125.8
CH6   233.0   99.670386  16.697368  64.6   88.7  105.3  114.0  119.9
CH7   233.0  107.186695  12.469806  88.3   96.3  104.6  117.6  140.3
CH8   233.0  109.668240  12.111580  86.5  100.5  105.9  118.2  144.4
CH9   233.0  109.457082  20.931974  73.8   94.5  103.6  132.6  150.8
CH10  233.0   99.805150  21.094146  53.6   90.7  103.2  116.6  134.4
CH11  233.0  100.838197  33.553354  43.1   67.8  103.1  128.2  155.3
CH12  233.0  123.141631  51.028257  49.7   77.8  114.6  164.3  223.2
CH14  233.0  110.590129  30.909061  60.3   87.5  105.7  139.2  163.1
LA0   233.0  108.613305  30.331889  55.4   86.5  109.6  135.6  155.1
LA1   233.0  107.760515  27.195640  59.0   90.4  104.7  131.3  158.5
LA2   233.0  105.712017  25.783185  56.8   85.0  109.2  129.3  143.5
LA3   233.0  114.969099  30.519721  68.0   91.8  107.3  137.9  170.1
LA4   233.0  113.193562  36.200918  50.3   86.8  114.4  149.5  161.7
LA5   233.0  102.063519  31.064868  50.6   74.6  108.5  126.1  151.5
LA6   233.0  101.574249  16.166241  67.4   92.2  104.2  113.8  124.8
LA7   233.0  108.037768  16.661455  79.7   95.2  105.5  123.5  141.7
LA8   233.0  111.380687   9.440714  91.5  104.1  110.7  117.1  132.6
LA9   233.0   96.687983  14.361771  67.4   90.2  100.4  105.7  122.1
LA10  233.0  101.493562  25.096327  51.8   84.7  105.9  121.6  139.3
LA11  233.0  121.043777  49.830553  42.6   90.4  120.4  168.2  218.1
LA12  233.0  122.622318  52.749634  46.4   75.2  114.6  163.6  220.9
LA14  233.0  106.381974  22.257080  68.9   88.7  108.1  124.1  141.8
NY0   233.0  110.480687  31.226034  59.7   84.6  109.6  138.4  161.8
NY1   233.0  110.278970  27.988656  62.8   90.9  105.6  139.1  157.7
...     ...         ...        ...   ...    ...    ...    ...    ...
NY7   233.0  101.759657  10.218290  84.0   94.1  101.4  111.2  120.0
NY8   233.0  109.725322  11.903484  91.1  100.5  105.7  119.1  146.7
NY9   233.0  104.506438  18.587874  69.1   93.6  100.8  120.9  139.7
NY10  233.0  101.221459  23.259601  54.9   87.8  104.2  116.2  140.6
NY11  233.0  109.711159  36.800570  55.4   75.8  112.1  139.9  171.9
NY12  233.0  127.378112  52.927450  54.5   79.0  117.2  171.8  226.0
NY14  233.0  110.133047  28.867370  62.5   83.0  110.4  135.8  158.6
PH0   233.0  109.332189  29.965311  58.5   86.3  109.6  136.3  158.0
PH1   233.0  109.226609  25.671544  64.7   92.8  105.7  135.9  156.3
PH2   233.0  101.015880  23.846312  53.5   83.4  106.1  122.9  133.9
PH3   233.0  112.168670  32.096056  65.2   87.4  105.8  138.4  168.5
PH4   233.0  120.271245  40.154271  60.7   87.9  115.1  157.9  184.2
PH5   233.0   94.530472  23.840442  45.1   80.0  101.2  112.8  125.9
PH6   233.0   98.823176  16.242786  63.4   88.8  102.0  112.3  120.9
PH7   233.0  100.781974   7.999953  82.8   94.4  101.0  106.6  120.6
PH8   233.0   94.796567  10.798978  62.1   87.6   96.0  102.5  120.2
PH9   233.0  101.418884  12.649584  75.6   93.7  103.5  110.1  130.1
PH10  233.0  103.704721  26.299774  52.1   87.2  106.2  123.3  146.2
PH11  233.0  113.978112  41.358808  47.1   83.4  116.0  157.2  179.9
PH12  233.0  126.609442  53.920057  51.7   78.3  121.0  169.2  229.8
PH14  233.0  110.206009  27.726580  67.8   85.2  105.7  133.3  165.0
DA1   233.0  104.640343  23.062150  59.7   90.5  105.2  127.6  137.5
DT1   233.0  105.102575  23.074736  61.5   92.0  102.9  127.6  142.7
HS1   233.0  107.103863  26.126698  60.8   89.4  105.0  133.5  149.3
PI1   233.0  103.897854  21.581720  63.1   92.3  101.4  122.7  142.8
SF1   233.0  109.288412  28.032423  60.1   91.0  104.9  136.6  159.7
BA1 

In [13]:
# Show the NaN values for the city CPI and for food-at-home. 
usa.loc[:, ['Year-Month', 'DA0', 'DA1', 'DA2', 'DA3', 'DT0', 'DT1', 'HS0', 'HS1', 'PI0', 'PI1', 'SF0', 'SF1', 'BA0', 'BA1', 'BO0', 'BO1', 'ST0', 'ST1', 'WA0', 'WA1']]

Year-Month    DA0    DA1    DA2    DA3    DT0    DT1    HS0    HS1    PI0  \
0   1976-01-01    NaN   61.1   52.9    NaN   55.7   63.0   53.7   61.7   53.8   
1   1976-02-01   52.2   60.6   54.0   61.1   55.8   62.7    NaN   61.5    NaN   
2   1976-03-01    NaN   59.7   54.4    NaN   55.7   61.7    NaN   61.0    NaN   
3   1976-04-01    NaN   60.0   54.5    NaN   56.0   62.1   54.3   60.8   54.6   
4   1976-05-01   53.0   60.2   54.9   60.7   56.3   62.3    NaN   61.2    NaN   
5   1976-06-01    NaN   60.7   55.0    NaN   56.5   62.3    NaN   61.4    NaN   
6   1976-07-01    NaN   60.9   55.3    NaN   56.9   62.7   55.2   62.0   55.2   
7   1976-08-01   53.9   60.9   55.6   61.5   57.1   62.3    NaN   61.8    NaN   
8   1976-09-01    NaN   60.2   56.0    NaN   57.6   61.8    NaN   61.6    NaN   
9   1976-10-01    NaN   60.3   56.3    NaN   57.7   61.5   56.7   61.9   56.0   
10  1976-11-01   54.7   60.3   56.7   61.9   58.0   62.0    NaN   60.8    NaN   
11  1976-12-01    NaN   61.1   56.7    NaN   58.2   62.7    NaN   61.1    NaN   
12  1977-01-01    NaN   61.8   57.3    NaN   58.5   63.0   57.2   62.0   56.7   
13  1977-02-01   55.9   64.1   58.3   62.6   58.9   64.4    NaN   63.5    NaN   
14  1977-03-01    NaN   64.1   58.7    NaN   59.5   64.2    NaN   64.2    NaN   
15  1977-04-01    NaN   65.0   59.5    NaN   60.2   65.4   58.8   64.3   58.4   
16  1977-05-01   57.2   65.0   60.6   62.6   60.3   65.5    NaN   64.6    NaN   
17  1977-06-01    NaN   65.5   61.0    NaN   60.9   66.8    NaN   64.6    NaN   
18  1977-07-01    NaN   65.6   61.4    NaN   61.4   67.2   59.7   65.1   59.2   
19  1977-08-01   58.1   66.0   62.0   63.5   61.4   67.2    NaN   65.6    NaN   
20  1977-09-01    NaN   65.8   62.2    NaN   61.5   66.5    NaN   64.8    NaN   
21  1977-10-01    NaN   65.4   62.1    NaN   61.6   66.5   60.1   64.5   60.1   
22  1977-11-01   58.6   66.4   62.5   64.4   62.1   67.3    NaN   65.3    NaN   
23  1977-12-01    NaN   66.1   62.7    NaN   62.0   67.3    NaN   65.4    NaN   
24  1978-01-01    NaN   66.9   63.1    NaN   62.2   68.4   61.3   66.2   60.6   
25  1978-02-01   59.5   68.1   63.5   65.5   62.4   69.4    NaN   67.0    NaN   
26  1978-03-01    NaN   68.8    NaN    NaN   63.4   70.5    NaN   68.2    NaN   
27  1978-04-01   60.3   69.7   63.5   67.6   64.0   71.9   63.3   69.2   62.3   
28  1978-05-01    NaN   70.7    NaN    NaN   64.7   72.9    NaN   70.2    NaN   
29  1978-06-01   61.7   72.3   64.9   68.1   65.4   74.3   64.7   71.7   63.5   
..         ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
203 1992-12-01  134.6  128.9  149.2  151.7  137.1  134.4  129.3  133.9  137.3   
204 1993-01-01    NaN  129.8    NaN    NaN    NaN  134.2    NaN  132.9    NaN   
205 1993-02-01  135.4  129.0  149.3  153.7  138.3  134.1  131.7  136.5  139.2   
206 1993-03-01    NaN  129.3    NaN    NaN    NaN  134.3    NaN  136.4    NaN   
207 1993-04-01  137.0  131.1  149.7  152.9  138.7  134.0  131.8  130.2  139.6   
208 1993-05-01    NaN  131.6    NaN    NaN    NaN  135.9    NaN  133.2    NaN   
209 1993-06-01  136.2  130.6  149.5  154.7  139.1  134.4  132.9  132.8  139.5   
210 1993-07-01    NaN  126.6    NaN    NaN    NaN  134.4    NaN  136.0    NaN   
211 1993-08-01  138.1  128.6  150.2  153.9  139.9  134.8  133.0  138.2  140.4   
212 1993-09-01    NaN  128.1    NaN    NaN    NaN  135.9    NaN  136.6    NaN   
213 1993-10-01  139.6  127.4  150.2  154.4  141.9  136.6  136.6  140.4  140.6   
214 1993-11-01    NaN  130.9    NaN    NaN    NaN  136.9    NaN  141.2    NaN   
215 1993-12-01  138.8  133.2  150.6  170.4  140.2  136.3  136.5  143.2  141.1   
216 1994-01-01    NaN  133.5    NaN    NaN    NaN  136.7    NaN  145.8    NaN   
217 1994-02-01  139.2  130.6  150.1  163.6  141.7  136.4  137.0  142.5  142.6   
218 1994-03-01    NaN  131.9    NaN    NaN    NaN  136.3    NaN  143.9    NaN   
219 1994-04-01  140.3  133.0  150.0  163.2  142.6  137.0  136.8  144.4  143.9   
220 1994-05-01    NaN  130.7    NaN

For Dallas (DA), there's data for DA1 (food at home) for all months.  But, for DA0 (city CPI), there is data for one out of every three months through 1977 and then there's data for every other month.  Similar observations for other cities such as Baltimore (BA), Boston (BO), Detroit (DT), Houston (HS), Pittsburg (PI), San Francisco (SF), St. Louis (ST), and Washington D.C. (WA).  It's odd that there's more data for food-at-home than for city-wide CPI.  It makes one wonder if the city-wide column may be inadvertantly switched to the one-numbered column.

In [14]:
# 12. Medical Care
usadesc.loc[[f'{c}12' for c in cities['usa'].keys()]]

count        mean        std   min      25%     50%      75%    max
CH12  233.0  123.141631  51.028257  49.7   77.800  114.60  164.300  223.2
LA12  233.0  122.622318  52.749634  46.4   75.200  114.60  163.600  220.9
NY12  233.0  127.378112  52.927450  54.5   79.000  117.20  171.800  226.0
PH12  233.0  126.609442  53.920057  51.7   78.300  121.00  169.200  229.8
DA12  112.0  124.053571  49.715234  48.5   78.450  122.90  164.600  216.2
DT12  182.0  109.876374  41.935942  54.1   75.125   98.40  134.750  208.8
HS12  112.0  124.052679  50.270178  45.5   79.550  120.15  164.925  217.3
PI12  112.0  122.806250  49.584200  45.8   77.975  117.80  163.075  211.6
SF12  163.0  138.006135  47.314456  51.7  101.400  138.10  179.750  209.2
BA12  112.0  126.035714  48.736750  56.0   84.900  118.95  162.200  231.0
BO12  113.0  137.841593  63.850766  51.6   82.300  126.50  182.700  265.1
MI12  105.0  123.830476  40.555252  59.3   92.000  120.10  159.700  201.7
ST12  112.0  122.355357  47.352182  49.1   81.500  117.90  160.950  210.6
WA12  113.0  123.838938  48.171971  51.7   78.400  119.40  164.300  209.3

In general, medical care costs have risen more over the years than costs for other categories--CPI for this category is above 200 in 1995 for all cities, whereas CPI for most other categories is below 150.  Medical care costs in Boston (BO) have risen over 265.0 in 1995!!

In [15]:
cols14 = [f'{c}14' for c in cities['usa']]
# usa.loc[:11, ['Unnamed: 0'] + cols11]  ## 1976
# usa.loc[228:, ['Unnamed: 0'] + cols11]  ## 1995
# pri
# max(x for x in usa.loc[228:, ['Unnamed: 0'] + cols13].max()[1:] if x)
usa.loc[:, ['Year-Month'] + cols14]

Year-Month   CH14   LA14   NY14   PH14   DA14   DT14   HS14   PI14   SF14  \
0   1976-01-01   60.3   68.9   62.5   68.7    NaN   72.1   59.9   64.0    NaN   
1   1976-02-01   60.5   69.4   62.9   68.1   65.3   72.4    NaN    NaN    NaN   
2   1976-03-01   61.3   69.6   63.0   67.8    NaN   72.9    NaN    NaN   63.8   
3   1976-04-01   61.6   70.1   63.3   68.4    NaN   73.5   60.6   64.4    NaN   
4   1976-05-01   61.8   70.2   63.7   68.8   67.2   73.6    NaN    NaN    NaN   
5   1976-06-01   61.6   70.2   64.8   68.9    NaN   73.0    NaN    NaN   64.5   
6   1976-07-01   61.8   71.2   65.0   69.2    NaN   73.2   61.7   64.9    NaN   
7   1976-08-01   61.9   70.8   65.2   69.2   67.8   73.0    NaN    NaN    NaN   
8   1976-09-01   62.0   71.1   65.7   68.6    NaN   74.2    NaN    NaN   65.6   
9   1976-10-01   62.5   70.7   66.1   69.0    NaN   74.3   61.8   65.7    NaN   
10  1976-11-01   62.8   71.8   65.9   69.2   68.5   74.5    NaN    NaN    NaN   
11  1976-12-01   62.8   72.0   66.3   69.8    NaN   74.8    NaN    NaN   66.2   
12  1977-01-01   63.8   72.1   66.5   70.4    NaN   74.9   63.1   66.7    NaN   
13  1977-02-01   63.8   73.1   67.1   70.4   68.5   74.9    NaN    NaN    NaN   
14  1977-03-01   64.5   73.2   66.7   70.5    NaN   75.1    NaN    NaN   66.7   
15  1977-04-01   63.9   73.4   66.9   71.1    NaN   75.0   63.3   67.5    NaN   
16  1977-05-01   64.3   72.7   67.3   71.2   69.7   75.2    NaN    NaN    NaN   
17  1977-06-01   64.3   72.8   69.1   72.0    NaN   75.2    NaN    NaN   67.4   
18  1977-07-01   64.2   72.9   69.1   72.0    NaN   75.4   64.5   68.6    NaN   
19  1977-08-01   64.6   73.6   69.3   71.7   70.8   75.7    NaN    NaN    NaN   
20  1977-09-01   64.6   73.5   70.4   71.9    NaN   76.5    NaN    NaN   68.2   
21  1977-10-01   65.2   73.3   70.4   72.4    NaN   76.3   64.4   69.4    NaN   
22  1977-11-01   65.7   73.4   69.4   72.8   69.1   76.8    NaN    NaN    NaN   
23  1977-12-01   66.0   73.8   69.5   73.2    NaN   77.3    NaN    NaN   68.8   
24  1978-01-01   66.7   71.9   69.9   73.5    NaN   77.5   64.8   70.4    NaN   
25  1978-02-01   67.0   73.8   71.2   73.7   68.8   77.7    NaN    NaN   66.6   
26  1978-03-01   67.2   74.6   71.9   73.5    NaN   78.3    NaN    NaN    NaN   
27  1978-04-01   67.4   76.3   72.5   74.2   68.4   79.9   65.8   71.1   70.2   
28  1978-05-01   67.4   76.0   72.5   74.3    NaN   79.9    NaN    NaN    NaN   
29  1978-06-01   67.2   74.3   72.1   75.6   70.5   79.5   64.5   71.4   71.2   
..         ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
203 1992-12-01  150.3  135.1  148.5  144.6  143.4  131.4  149.8  146.9  158.1   
204 1993-01-01  151.0  135.9  149.4  143.2    NaN    NaN    NaN    NaN  155.5   
205 1993-02-01  152.7  135.5  149.0  134.2  144.1  134.0  152.9  147.4  160.9   
206 1993-03-01  152.9  135.7  149.5  140.7    NaN    NaN    NaN    NaN  161.3   
207 1993-04-01  153.9  135.4  149.9  150.6  142.6  136.2  152.2  148.4  161.7   
208 1993-05-01  152.4  134.1  149.1  149.6    NaN    NaN    NaN    NaN  163.5   
209 1993-06-01  155.2  133.0  149.1  149.5  143.5  137.5  153.4  151.0  163.1   
210 1993-07-01  153.9  133.9  148.7  149.5    NaN    NaN    NaN    NaN  162.8   
211 1993-08-01  153.4  135.0  150.2  149.7  143.7  138.6  150.8  152.0  163.1   
212 1993-09-01  155.2  136.0  150.5  150.3    NaN    NaN    NaN    NaN  162.9   
213 1993-10-01  154.7  137.7  151.4  151.1  145.8  140.4  158.7  150.7  163.6   
214 1993-11-01  154.8  138.9  151.2  150.6    NaN    NaN    NaN    NaN  164.3   
215 1993-12-01  156.6  138.8  149.8  150.9  145.1  142.0  151.8  150.3  164.6   
216 1994-01-01  158.4  139.4  151.5  155.7    NaN    NaN    NaN    NaN  164.0   
217 1994-02-01  157.2  138.0  152.9  157.4  147.1  143.8  156.6  150.7  164.9   
218 1994-03-01  159.0  138.1  154.1  158.3    NaN    NaN    NaN    NaN  166.6   
219 1994-04-01  159.6  138.5  152.7  163.8  145.6  145.5  155.7  152.2  165.5   
220 1994-05-01  160.3  138.3  153.0